### PR TITLE
Use keys in map component - second try

### PR DESCRIPTION
**Pull request by AdamRamberg**
This PR makes #56 obsolete. 

Today the `<Map />` component is internally using the array index as a key when rendering out each item. This causes a similar problem to what is discussed in the [official React docs](https://reactjs.org/docs/lists-and-keys.html#keys). This PR introduces a prop called `keySelector`, which enables the consumer to select a unique key from the item data. If no [keySelector](url) is provided we default back to index as key.

In this PR we kept both implementations to be able to do a side-by-side comparison of the Map component: <Map /> (old) and `<MapWithKey />` (new)

# Benchmarking
When comparing the new implementation with the old one it seems to be quite a lot faster (36.7%).
### `yarn compare listMemoFacet listMemoState 70`
Iteration 0:    267302  509240  52.49%
Iteration 1:    245532  450027  54.56%
Iteration 2:    263251  440437  59.77%
Iteration 3:    232783  437802  53.17%
Iteration 4:    246622  425732  57.93%
Iteration 5:    219669  428715  51.24%
Iteration 6:    201011  459167  43.78%
Iteration 7:    259807  412293  63.02%
Iteration 8:    238269  436092  54.64%
Iteration 9:    283545  490877  57.76%
Relative performance listMemoFacet/listMemoState: 54.73

### `yarn compare listMemoWithKeyFacet listMemoState 70`
Iteration 0:    147970  415106  35.65%
Iteration 1:    136320  440923  30.92%
Iteration 2:    142283  417188  34.11%
Iteration 3:    151806  408697  37.14%
Iteration 4:    144983  427303  33.93%
Iteration 5:    139096  404501  34.39%
Iteration 6:    151532  429485  35.28%
Iteration 7:    129735  399111  32.51%
Iteration 8:    148728  412292  36.07%
Iteration 9:    146486  395972  36.99%
Relative performance listMemoWithKeyFacet/listMemoState: 34.67

# Discussion
- Should the `keySelector` prop be mandatory or optional? The best practice is not to use the array index as a key, but making it mandatory would be a breaking change
- Since this is a bug that only appears when adding/removing items that are not at end of the array (not that common use case), should we have two implementations? One where the array length is static (ie. could use the index as a key) and one dynamic

### CODE REVIEW

**Review by vb — COMMENTED** [review]
Nice that you have revisited this! Great work!
Overall I think it looks great.
Two things:
- I think we need to make it clearer that the keys need to be unique.
- Add unit tests for the use cases this solves (adding/removing from the start of an array etc)

File: examples/benchmarking/src/listMemoWithKeyFacet.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      previousIndex = current.index
+      return false
+    }
+
+    return true
+  }
+}
+
+const dataKeySelector = (data: Data) => {
+  return data.name
```
[the remark this answers is not in the document]
Comment (AdamRamberg) [reply examples/benchmarking/src/listMemoWithKeyFacet.tsx:106]: Good point! I will make it unique for the sake of the test 👍

**Review by Shenato — APPROVED** [review]
Great Addition, I think if there's no gotchas we should definitely replace the old Map component, make `keySelector` optional and default to indices

**Review by pirelenito — CHANGES_REQUESTED** [review]
I had a chat with Adam over Slack, and here is a summary of it:

- We are not removing old keys from the `itemFacetsMap` as the array changes. So this will be a memory leak

- Given this spread `{...childProps}` is only actually two props, its possible that maybe setting them explicitly is better from a readability and perf perspective

- Although understandable the reasoning to have the useState properly initialised on mount, its also possible that the extra complexity is not worth it. Also, it might be causing a double update anyway.
  - First when you mount and you have the facets with their “innitial values”
  - Then when the `useFacetEffect` runs and actually sets the facets again with their “innitial values”. This would only be free, if people passed the `equalityCheck`
 
- Won’t this cause issues `mapFacetSingleLightweight(array, (a) => a[i])` when something moves within the array? (not sure tbh)

- This needs lots of unit tests, as it can have quite weird edge-cases

- Performance:
  - it will be interesting to compare the scenario where this would actually make a difference: having an array that you are swapping the position of two items. There is a hunch that the new implementation will actually loose, when compared to the current one (as it reconciles) :see_no_evil:
  - Also would be worth comparing head-to-head using the benchmark utility (ex: `yarn compare markerWithKeyFacet markerFacet 100 19000`)
